File: src/common/metrics.zig
/// metrics.zig — Generic latency histogram primitive (M8.1)
///
/// Thread-safe Prometheus-style histogram for latency measurement.
/// Used by coral (cache tier resolution), vector (HNSW/DB queries),
/// and llm (HTTP call duration) subsystems.
///
/// Bucket boundaries (milliseconds): 1, 5, 10, 25, 50, 100, 250, 500, 1000,
/// 2500, 5000 — matches Prometheus histogram standard boundaries for HTTP SLOs.
const std = @import("std");

// ---------------------------------------------------------------------------
// Histogram bucket boundaries
// ---------------------------------------------------------------------------

/// Upper-bound millisecond thresholds for each finite bucket.
/// The last implicit bucket is +Inf (all observations).
pub const BUCKET_MS = [_]u64{ 1, 5, 10, 25, 50, 100, 250, 500, 1_000, 2_500, 5_000 };
pub const BUCKET_COUNT: usize = BUCKET_MS.len + 1; // +1 for the +Inf bucket

// ---------------------------------------------------------------------------
// LatencyHistogram
// ---------------------------------------------------------------------------

/// Thread-safe latency histogram with atomic bucket counters.
/// Tracks latency distributions; managed by owner; key invariant is accurate measurement.
pub const LatencyHistogram = struct {
    /// Bucket counts: index i accumulates observations where
    /// BUCKET_MS[i-1] < latency_ms ≤ BUCKET_MS[i].
    /// Index BUCKET_COUNT-1 is the +Inf bucket.
    buckets: [BUCKET_COUNT]std.atomic.Value(u64) = undefined,
    /// Total count of all observations.
    count: std.atomic.Value(u64) = std.atomic.Value(u64).init(0),
    /// Sum of all observed latencies in milliseconds.
    sum_ms: std.atomic.Value(u64) = std.atomic.Value(u64).init(0),

    pub fn init() LatencyHistogram {
        var h = LatencyHistogram{};
        for (&h.buckets) |*b| b.* = std.atomic.Value(u64).init(0);
        return h;
    }

    /// Record one observation of `latency_ms` milliseconds.
    pub fn observe(self: *LatencyHistogram, latency_ms: u64) void {
        _ = self.count.fetchAdd(1, .monotonic);
        _ = self.sum_ms.fetchAdd(latency_ms, .monotonic);
        // Find the first bucket whose upper bound ≥ latency_ms.
        for (BUCKET_MS, 0..) |bound, i| {
            if (latency_ms <= bound) {
                _ = self.buckets[i].fetchAdd(1, .monotonic);
                return;
            }
        }
        // Falls into the +Inf bucket.
        _ = self.buckets[BUCKET_COUNT - 1].fetchAdd(1, .monotonic);
    }

    /// Return the cumulative count for bucket `idx` (Prometheus-style:
    /// bucket[i] counts all observations ≤ BUCKET_MS[i]).
    pub fn cumulativeBucket(self: *const LatencyHistogram, idx: usize) u64 {
        var cum: u64 = 0;
        var i: usize = 0;
        while (i <= idx) : (i += 1) {
            cum += self.buckets[i].load(.monotonic);
        }
        return cum;
    }

    /// Estimate the `p`-th percentile latency (0.0–1.0) from histogram buckets.
    /// Returns the upper bound of the first bucket whose cumulative count
    /// reaches or exceeds `p * total`.  Returns 0 when no observations.
    pub fn estimatePercentile(self: *const LatencyHistogram, p: f64) u64 {
        const total = self.count.load(.monotonic);
        if (total == 0) return 0;
        const target: u64 = @intFromFloat(@ceil(p * @as(f64, @floatFromInt(total))));
        var cum: u64 = 0;
        for (BUCKET_MS, 0..) |bound, i| {
            cum += self.buckets[i].load(.monotonic);
            if (cum >= target) return bound;
        }
        // All observations fall in the +Inf bucket.
        return std.math.maxInt(u64);
    }
};

// =============================================================================
// Tests
// =============================================================================

const testing = std.testing;

test "LatencyHistogram: observe increments count and sum" {
    var h = LatencyHistogram.init();
    h.observe(10);
    h.observe(20);
    try testing.expectEqual(@as(u64, 2), h.count.load(.monotonic));
    try testing.expectEqual(@as(u64, 30), h.sum_ms.load(.monotonic));
}

test "LatencyHistogram: observe routes to correct bucket" {
    var h = LatencyHistogram.init();
    h.observe(1); // bucket 0 (≤1ms)
    h.observe(5); // bucket 1 (≤5ms)
    h.observe(6); // bucket 2 (≤10ms)
    try testing.expectEqual(@as(u64, 1), h.buckets[0].load(.monotonic));
    try testing.expectEqual(@as(u64, 1), h.buckets[1].load(.monotonic));
    try testing.expectEqual(@as(u64, 1), h.buckets[2].load(.monotonic));
}

test "LatencyHistogram: large value goes to +Inf bucket" {
    var h = LatencyHistogram.init();
    h.observe(100_000);
    try testing.expectEqual(@as(u64, 1), h.buckets[BUCKET_COUNT - 1].load(.monotonic));
}

test "LatencyHistogram: cumulative bucket includes earlier buckets" {
    var h = LatencyHistogram.init();
    h.observe(1); // bucket 0
    h.observe(5); // bucket 1
    // cumulativeBucket(1) = bucket[0] + bucket[1] = 2
    try testing.expectEqual(@as(u64, 2), h.cumulativeBucket(1));
}

test "LatencyHistogram: estimatePercentile returns 0 when empty" {
    const h = LatencyHistogram.init();
    try testing.expectEqual(@as(u64, 0), h.estimatePercentile(0.5));
}

test "LatencyHistogram: estimatePercentile p50 with known observations" {
    var h = LatencyHistogram.init();
    h.observe(1);
    h.observe(1);
    h.observe(10);
    h.observe(10);
    try testing.expectEqual(@as(u64, 1), h.estimatePercentile(0.5));
    try testing.expectEqual(@as(u64, 10), h.estimatePercentile(0.75));
}

test "LatencyHistogram: thread-safe concurrent observe" {
    var h = LatencyHistogram.init();
    const N_THREADS = 4;
    const N_OBS = 100;

    const S = struct {
        fn worker(hist: *LatencyHistogram) void {
            for (0..N_OBS) |_| hist.observe(5);
        }
    };

    var threads: [N_THREADS]std.Thread = undefined;
    for (&threads) |*t| t.* = try std.Thread.spawn(.{}, S.worker, .{&h});
    for (&threads) |*t| t.join();

    try testing.expectEqual(@as(u64, N_THREADS * N_OBS), h.count.load(.monotonic));
}

File: src/common/src/metrics.rs
use std::sync::atomic::{AtomicU64, Ordering};
use std::time::Instant;

pub const BUCKET_MS: [u64; 11] = [1, 5, 10, 25, 50, 100, 250, 500, 1000, 2500, 5000];
pub const BUCKET_COUNT: usize = 12;

pub struct LatencyHistogram {
    buckets: [AtomicU64; BUCKET_COUNT],
    count: AtomicU64,
    sum: AtomicU64,
}

impl LatencyHistogram {
    pub fn new() -> Self {
        Self {
            buckets: Default::default(),
            count: AtomicU64::new(0),
            sum: AtomicU64::new(0),
        }
    }

    fn bucket_index(duration_ms: u64) -> usize {
        for (i, &bound) in BUCKET_MS.iter().enumerate() {
            if duration_ms <= bound {
                return i;
            }
        }
        BUCKET_COUNT - 1
    }

    pub fn observe(&self, duration_ms: u64) {
        let idx = Self::bucket_index(duration_ms);
        self.buckets[idx].fetch_add(1, Ordering::Relaxed);
        self.count.fetch_add(1, Ordering::Relaxed);
        self.sum.fetch_add(duration_ms, Ordering::Relaxed);
    }

    pub fn observe_duration(&self, start: Instant) {
        let elapsed = start.elapsed();
        self.observe(elapsed.as_millis() as u64);
    }

    pub fn count(&self) -> u64 {
        self.count.load(Ordering::Relaxed)
    }

    pub fn sum_ms(&self) -> u64 {
        self.sum.load(Ordering::Relaxed)
    }

    pub fn bucket(&self, idx: usize) -> u64 {
        if idx < BUCKET_COUNT {
            self.buckets[idx].load(Ordering::Relaxed)
        } else {
            0
        }
    }

    pub fn estimate_percentile(&self, pct: f64) -> u64 {
        let total = self.count();
        if total == 0 {
            return 0;
        }
        let target = (total as f64 * pct / 100.0) as u64;
        let mut cumulative = 0u64;
        for (i, &bound) in BUCKET_MS.iter().enumerate() {
            cumulative += self.buckets[i].load(Ordering::Relaxed);
            if cumulative >= target {
                return bound;
            }
        }
        *BUCKET_MS.last().unwrap_or(&5000)
    }
}

impl Default for LatencyHistogram {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::sync::Arc;
    use std::thread;

    #[test]
    fn observe_increments_count_and_sum() {
        let h = LatencyHistogram::new();
        h.observe(10);
        assert_eq!(h.count(), 1);
        assert_eq!(h.sum_ms(), 10);
    }

    #[test]
    fn observe_routes_to_correct_bucket() {
        let h = LatencyHistogram::new();
        h.observe(1);
        h.observe(10);
        h.observe(100);
        assert_eq!(h.bucket(0), 1);  // 1ms ≤ 1 → bucket 0
        assert_eq!(h.bucket(2), 1);  // 10ms ≤ 10 → bucket 2
        assert_eq!(h.bucket(5), 1);  // 100ms ≤ 100 → bucket 5
    }

    #[test]
    fn large_value_goes_to_inf_bucket() {
        let h = LatencyHistogram::new();
        h.observe(99999);
        assert_eq!(h.bucket(BUCKET_COUNT - 1), 1);
    }

    #[test]
    fn estimate_percentile_returns_zero_when_empty() {
        let h = LatencyHistogram::new();
        assert_eq!(h.estimate_percentile(50.0), 0);
    }

    #[test]
    fn estimate_percentile_p50_with_known() {
        let h = LatencyHistogram::new();
        h.observe(1);
        h.observe(10);
        h.observe(100);
        let p50 = h.estimate_percentile(50.0);
        assert!(p50 <= 100);
    }

    #[test]
    fn cumulative_bucket_includes_earlier() {
        let h = LatencyHistogram::new();
        h.observe(5);
        h.observe(50);
        assert_eq!(h.bucket(0), 0);  // No value ≤ 1ms
        assert_eq!(h.bucket(1), 1);  // 5ms ≤ 5 → bucket 1
        assert_eq!(h.bucket(4), 1);  // 50ms ≤ 50 → bucket 4
    }

    #[test]
    fn thread_safe_concurrent_observe() {
        let h = Arc::new(LatencyHistogram::new());
        let mut handles = Vec::new();
        for _ in 0..10 {
            let h_clone = Arc::clone(&h);
            handles.push(thread::spawn(move || {
                for _ in 0..100 {
                    h_clone.observe(5);
                }
            }));
        }
        for handle in handles {
            handle.join().unwrap();
        }
        assert_eq!(h.count(), 1000);
    }

    #[test]
    fn observe_duration_records_millis() {
        let h = LatencyHistogram::new();
        let start = Instant::now();
        std::thread::sleep(std::time::Duration::from_millis(1));
        h.observe_duration(start);
        assert!(h.count() >= 1);
        assert!(h.sum_ms() >= 1);
    }

    #[test]
    fn bucket_out_of_range_returns_zero() {
        let h = LatencyHistogram::new();
        assert_eq!(h.bucket(99), 0);
    }

    #[test]
    fn estimate_percentile_returns_max_when_target_in_last_bucket() {
        let h = LatencyHistogram::new();
        h.observe(99999);
        let pct = h.estimate_percentile(100.0);
        assert!(pct >= 5000);
    }

    #[test]
    fn default_creates_empty_histogram() {
        let h = LatencyHistogram::default();
        assert_eq!(h.count(), 0);
    }
}
